### DISCUSSION TITLE
General question about multiple in-and output

### DISCUSSION THREAD

**Discussion by dezenn** [neuraloperator/neuraloperator#346] [Announcements]
I would like to use the Fourier neural operator to train on time-dependent functions. But I have several input function features ( like e.g. voltage and power, both dependent on time) and several output function features. The way these features are discretized is the same for all features and datapoints.

Is the way the FNO is implemented here useable for such a scenario? I see that there exist the arguments input_channel and output_channel but I am not sure if I can use these because this could also mean that I can 'only' learn on one feature but that one feature is e.g. 3-dimensional.

And as far as I understand FNOs are not suitable when taking different input types, for example besides input functions I would additionally consider a vector of scalars as input, or am I mistaken?

Thanks in advance
Best regards
Dean

**Comment by gokhalen** [comment]
@dezenn 

I believe what you are asking for is possible - please take a look at 

https://github.com/neuraloperator/neuraloperator/blob/master/fourier_3d.py

That's an old example - but useful for learning.  There they map time dependent velocity fields into the future. 

-Nachiket

**Comment by gokhalen** [comment]
Also see the following reply by @zongyi-li 

https://github.com/neuraloperator/neuraloperator/issues/48#issuecomment-886083377

And also https://github.com/neuraloperator/neuraloperator/issues/47

**Comment by dezenn** [comment]
@gokhalen Thank you very much for your answer. I really appreciate it. 

As far as I understand it now, I can not use the FNO for my scenario. What they do is that they map 10 timesteps of a 3d velocity field to another timestep. But what I want to learn is an operator $\mathcal{A}:\mathcal{C}([0,T])^n\times\mathbb{R}^l\rightarrow\mathcal{C}([0,T])^m$ with $n,m,l\in\mathbb{N}$ and $T\in\mathbb{R}$. So I have a cartesian product of the space of continuous functions and in each space I have a function describing the time-dependent evolution of a physical entity.

**Comment by gokhalen** [comment]
@dezenn  Sorry if I misunderstood, but it seems to me that your operator maps a function of $n$ timelike variables and $l$ space-like variabes to a function of $m$ time-like variables. Seems to me that this is possible.

Edit: Now that I understand your question a little better, you might need to perform FFTs on higher dimensional grids.

**Comment by dezenn** [comment]
@gokhalen No problem. I will look into FFTs regarding higher dimensional grids.

Note that the $l$ variables are not space-depenndent. You can think of them as global parameters that are constant during the temporal evolution but vary from data point to data point, or from operator evaluation to operator evaluation.

To understand my setting better: the extension MIONet of the network architecture DeepONet is applicable for my case. But the practical implementation of the MIONet in the Python library DeepXDE has some limitations. So I wondered if the FNO is suitable for my case since I also think that the FFT would perform well on my type of data.

**Reply by nickhnelsen** [reply]
To handle finite-dimensional inputs or outputs (like the $l$ variables in your example) with FNO-like architectures, take a look at:
https://github.com/nickhnelsen/fourier-neural-mappings

I think the above code can be adapted to your problem. Alternatively, you can treat the $l$ variables as the values of $l$ constant functions. This increases your input space channel dimension to $n+l$.

**Reply by dezenn** [reply]
Thank you, I will definetely have a look at your fourier neural mappings!

So as far as I get it, usually the channels consist of a physical quantity and a grid, thus resulting in 2 input channels for the 1d case. In my case I have several physical quantities and they are all evaluated on the same points, therefore I do not need to know to incorporate the locations for my model. Nevertheless I have multiple channels as they are representing the number of my function features, right? So it is basically the same as I would use Conv1d layer.

**Reply by nickhnelsen** [reply]
Yes that sounds right! Since your physical quantities/"function features" are constant with respect to spatial location, you would would just broadcast the same constant value to fill the input tensor in the correct channel dimensions.
